src/main/java/org/elasticsearch/indices/SyncedFlushService.java

```diff
@@ -0,0 +1,755 @@
+        preSyncTimeout = settings.getAsTime(SETTING_PRE_SYNC_TIMEOUT, TimeValue.timeValueMinutes(5));
+        syncTimeout = settings.getAsTime(SETTING_SYNC_TIMEOUT, TimeValue.timeValueMinutes(5));
+        inflightOpsTimeout = settings.getAsTime(SETTING_IN_FLIGHT_OPS_TIMEOUT, TimeValue.timeValueMinutes(5));
+        indicesService.indicesLifecycle().addListener(new IndicesLifecycle.Listener() {
+            @Override
+            public void onShardInactive(final IndexShard indexShard) {
+                // we only want to call sync flush once, so only trigger it when we are on a primary
+                if (indexShard.routingEntry().primary()) {
+                    attemptSyncedFlush(indexShard.shardId(), new ActionListener<SyncedFlushResult>() {
+                        @Override
+                        public void onResponse(SyncedFlushResult syncedFlushResult) {
+                            logger.debug("{} sync flush on inactive shard returned successfully for sync_id: {}", syncedFlushResult.getShardId(), syncedFlushResult.syncId());
+                        }
+
+                        @Override
+                        public void onFailure(Throwable e) {
+                            logger.debug("{} sync flush on inactive shard failed", e, indexShard.shardId());
+                        }
+                    });
+                }
+            }
+        });
+    }
+
+    /*
+    * Tries to flush all copies of a shard and write a sync id to it.
+    * After a synced flush two shard copies may only contain the same sync id if they contain the same documents.
+    * To ensure this, synced flush works in three steps:
+    * 1. Flush all shard copies and gather the commit points for each copy after the flush
+    * 2. Ensure that there are no ongoing indexing operations on the primary
+    * 3. Perform an additional flush on each shard copy that writes the sync id
+    *
+    * Step 3 is only executed on a shard if
+    * a) the shard has no uncommitted changes since the last flush
+    * b) the last flush was the one executed in 1 (use the collected commit id to verify this)
+    *
+    * This alone is not enough to ensure that all copies contain the same documents. Without step 2 a sync id would be written for inconsistent copies in the following scenario:
+    *
+    * Write operation has completed on a primary and is being sent to replicas. The write request does not reach the replicas until sync flush is finished.
+    * Step 1 is executed. After the flush the commit points on primary contains a write operation that the replica does not have.
+    * Step 3 will be executed on primary and replica as well because there are no uncommitted changes on primary (the first flush committed them) and there are no uncommitted
+    * changes on the replica (the write operation has not reached the replica yet).
+    *
+    * Step 2 detects this scenario and fails the whole synced flush if a write operation is ongoing on the primary.
```

Comment:
can we add here that the neat part about this check in step 2. is that it will fail the flush on the primary even if there is an in-flight request coming into the primary right in between step 2. and 3. since  it will not execute the sync flush if there is a single modification in the buffer of these shards. I think that must be more clear here.